module/platform/Webots/src/Webots.cpp

```diff
@@ -0,0 +1,656 @@
+                // Halt and don't retry as the other end is clearly not Webots
+                close(fd);
+                powerplant.shutdown();
+            }
+        }
+        else {
+            // There was nothing sent
+            log<NUClear::DEBUG>("Connection was closed.");
+            setup_connection(server_address, port);
+            return;
+        }
+
+        // Set the real time of the connection initiation
+        connect_time = NUClear::clock::now();
+        // Reset the simulation connection time
+        utility::clock::last_update = std::chrono::steady_clock::now();
+
+        // Now that we are connected, we can set up our reaction handles with this file descriptor
+
+        // Receiving
+        read_io = on<IO>(fd, IO::READ).then([this]() {
+            // ************************** Receiving ***************************************
+            // Get the size of the message
+            uint32_t Nn;
+            if (recv(fd, &Nn, sizeof(Nn), 0) != sizeof(Nn)) {
+                log<NUClear::ERROR>("Failed to read message size from TCP connection");
+                return;
+            }
+
+            // Convert from network endian to host endian
+            const uint32_t Nh = ntohl(Nn);
+
+            // Get the message
+            std::vector<char> data(Nh, 0);
+            uint64_t message = uint64_t(recv(fd, data.data(), Nh, MSG_WAITALL));
+            if (message != Nh) {
+                log<NUClear::ERROR>("Failed to read message from TCP connection");
+                return;
+            }
+
+            log<NUClear::TRACE>("Received sensor measurements");
+
+            // Deserialise the message into a neutron
+            SensorMeasurements msg = NUClear::util::serialise::Serialise<SensorMeasurements>::deserialise(data);
+
+            translate_and_emit_sensor(msg);
+
+            // Service the watchdog
+            emit<Scope::WATCHDOG>(ServiceWatchdog<Webots>());
+
+            // ****************************** TIME **************************************
+
+            // Deal with time
+
+            // Save our previous deltas
+            const uint32_t prev_sim_delta  = sim_delta;
+            const uint32_t prev_real_delta = real_delta;
+
+            // Update our current deltas
+            sim_delta  = msg.time - current_sim_time;
+            real_delta = msg.real_time - current_real_time;
+
+            // Calculate our custom rtf - the ratio of the past two sim deltas and the past two real time deltas
+            utility::clock::custom_rtf =
+                static_cast<double>(sim_delta + prev_sim_delta) / static_cast<double>(real_delta + prev_real_delta);
+
+            // Update our current times
+            current_sim_time  = msg.time;
+            current_real_time = msg.real_time;
+        });
+
+        send_io =
+            on<IO, With<ServoTargets>, With<DarwinSensors>>(fd, IO::READ)
```

Comment:
`IO::READ`?? not `IO::WRITE`? Maybe this would be better as a 90Hz (or whatever) update loop like we have in in `HardwareIO`?

Reply:
90Hz is our current rate, but we will need to look at network bandwidth (probably more for images than sensors)